drivers/spi/spi_esp32_spim.c

```diff
@@ -0,0 +1,421 @@
+{
+	struct spi_esp32_data *data = DEV_DATA(dev);
+	spi_hal_context_t *hal = &(data->hal);
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		spi_hal_user_start(hal);
+		spi_context_update_tx(&data->ctx, data->dfs, 1);
+	}
+
+	k_timer_start(&spi_done_timer, K_MSEC(SPI_TIMEOUT_MSEC), K_NO_WAIT);
+	while (!spi_hal_usr_is_done(hal)) {
+		if (k_timer_status_get(&spi_done_timer) > 0) {
+			return -ETIMEDOUT;
+		}
+	}
+
+	if (spi_context_rx_buf_on(&data->ctx)) {
+		spi_hal_fetch_result(hal);
+		spi_context_update_rx(&data->ctx, data->dfs, 1);
+	}
+
+	return 0;
+}
+
+static int spi_esp32_init(const struct device *dev)
+{
+	const struct spi_esp32_config *cfg = DEV_CFG(dev);
+	struct spi_esp32_data *data = DEV_DATA(dev);
+
+	data->clock_dev = device_get_binding(cfg->clock_name);
+
+	spi_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static int spi_esp32_configure_pin(gpio_pin_t pin, int pin_sig,
+				   gpio_flags_t pin_mode)
+{
+	const char *device_name = gpio_esp32_get_gpio_for_pin(pin);
+	const struct device *gpio;
+	int ret;
+
+	if (!device_name) {
+		LOG_ERR("Could not find GPIO node on devicetree");
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		LOG_ERR("Could not bind to GPIO device");
+		return -EIO;
+	}
+
+	ret = gpio_pin_configure(gpio, pin, pin_mode);
+	if (ret < 0) {
+		LOG_ERR("SPI pin configuration failed");
+		return ret;
+	}
+
+	if (pin_mode == GPIO_INPUT) {
+		esp32_rom_gpio_matrix_in(pin, pin_sig, false);
+	} else {
+		esp32_rom_gpio_matrix_out(pin, pin_sig, false, false);
+	}
+
+	return 0;
+}
+
+static int spi_esp32_configure_cs_pin(const struct device *dev,
+				      const gpio_pin_t cs_pin)
+{
+	struct spi_esp32_data *data = DEV_DATA(dev);
+	const char *dev_name = gpio_esp32_get_gpio_for_pin(cs_pin);
+	const struct device *cs_gpio;
+
+	if (!dev_name) {
+		LOG_ERR("Could not find GPIO node on devicetree");
+		return -EINVAL;
+	}
+
+	cs_gpio = device_get_binding(dev_name);
+
+	if (!cs_gpio) {
+		LOG_ERR("Could not enable CS GPIO");
+		return -EIO;
+	}
+
+	const struct spi_cs_control cs_ctrl = { .gpio_dev = cs_gpio,
+						.delay = 0,
+						.gpio_pin = cs_pin,
+						.gpio_dt_flags =
+							GPIO_ACTIVE_LOW };
+
+	const struct spi_config cs_cfg = { .frequency =
+						   data->ctx.config->frequency,
+					   .operation =
+						   data->ctx.config->operation,
+					   .slave = data->ctx.config->slave,
+					   .cs = &cs_ctrl };
+
+	struct spi_context cs_ctx = {
+		.config = &cs_cfg,
+	};
+
+	memcpy(&data->ctx.config, &cs_ctx.config, sizeof(cs_ctx.config));
+
+	return 0;
+}
+
+static int spi_esp32_operation_to_mode(uint16_t operation)
+{
+	int cpol = SPI_MODE_GET(operation) & SPI_MODE_CPOL;
+	int cpha = SPI_MODE_GET(operation) & SPI_MODE_CPHA;
+
+	if (cpol == 0) {
+		if (cpha == 0) {
+			return 0;
+		}
+		return 1;
+	}
+
+	if (cpha == 0) {
+		return 2;
+	}
+	return 3;
+}
+
+static void spi_esp32_get_user_settings(struct spi_esp32_data *data)
+{
+#ifdef CONFIG_SPI_ESP32_HALFDUPLEX
```

Comment:
All options in this function should be configurable by device-tree properties instead of Kconfig option. They are instance specific, so device-tree will allow to configure each instance separately.